# UH-stat-bioinf
A collection of some of the projects I worked on during my master program in statistical bioinformatics. 

Purpose
--------
I created this repository to gather assignments and projects I made for my master studies, as well as some notes I made to reinforce new concepts and experiment with tools I encountered. It contains a wide variety of topics related to different courses, mostly related to statistics and bioinformatics. Some are just bare bones analyses or experiments, while others include a step-by-step description of the methods, objectives and interpretation. 

My aims throughout this venture were multifold. First, it gave me the chance to familiarise myself with the git version control system to efficiently track updates and share my results. Next, I intended to try out several tools used in data sciences: R, Python (and the SciPy package), and report and presentation methods (Knitr, Jupyter notebooks, markdown, \LaTeX).

Quick links
-----------

- [Tandem MS database search tool](MS2-database-search/): A python tool and Jupyter notebook documentation.
- [Connectivity scores](connectivity-scores/): Integration of connectivity scores of gene expression data and clustering on chemical structure of compounds. R markdown document and analysis.
- [RNA-seq differential expression analysis](RNA-Seq/): A Sweave/knitr R report comparing various R packages for differential expression analysis of RNA-seq data (edgeR, SAMSeq, EBSeq, voom+limma).
- [Machine learning - metabolic signature](CIMB/Metabolic%20signature.R): R analysis using machine learning methods to build a classifier based on metabolic data.
- [Analysis of differential gene expression](CIMB/): Some notes on multiple testing and bootstrap/permutation methods for micro-array expression data.
- [Bootstrap methods](CIM/): Some Jupyter notebooks showcasing how bootstrap methods work (hypothesis testing and confidence intervals).
