---
title: "Connectivity scores"
author: "Pieter Moris"
date: "21 februari 2016"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,cache = TRUE)
```

## Unraveling the connection between chemical structure and gene expression.

The aim of this analysis is to investigate the association between the structure of chemical compounds (or their predicted biological activity) and the effects they induce in a gene expression assay. We will group compounds with a similar structure (*chemical fingerprint*) or predicted bioactivity (*target prediction*) using clustering methods. Subsequently the *gene expression profiles* of the compounds within a cluster will be contrasted with each other using a connectivity score approach based on the work of [Lamb et al. (2006)](#lamb) and [Zhang & Gant (2008)](#zang) on the [Connectivity Map](https://www.broadinstitute.org/cmap/). 

Their methods enable us to contrast a number of query expression profiles with a set of reference profiles and assess the *connectivity* between compounds in terms of which genes are regulated up or down. The connection between the query and reference data can be both positive and negative. In the former case, highly up-regulated genes in one set will also be up-regulated in the other set, and vice versa for down-regulated genes. In the latter case, genes with a high expression in one assay, might be strongly down-regulated in the other one. Both situations imply that the two compounds interfere with the same biological processes. If two sets are weakly connected, there is no correlation or overlap between the top genes of either set.

*To do: description and origins of dataset.*

## Clustering chemical compounds
We will employ hierarchical clustering to group the chemical compounds based on their fingerprints or target predictions. In either case we are dealing with a 0/1 matrix where the rows correspond to different compounds. For the chemical fingerprints the columns indicate the presence or absence of specific chemical structures. Together this sequence of ones and zeros describes the entire molecular scaffolding of the compound. In essence we are dealing with a bit string for each compound.

*To do: add description of bioactivity matrix + describe dimensions of data.*

Let's take a look at the chemical fingerprints first. Here is a sneak peak at the data:

```{r,echo=FALSE}
load("MCF7Data.RData")
head(fingerprintMat[,1:4])
```



## References
<a name="lamb"></a>- Lamb, J. 2006. "The Connectivity Map: Using Gene-Expression Signatures To Connect Small Molecules, Genes, And Disease". Science 313 (5795): 1929-1935. doi:10.1126/science.1132939.

<a name="zang"></a> - Zhang, Shu-Dong, and Timothy W Gant. 2008. "A Simple And Robust Method For Connecting Small-Molecule Drugs Using Gene-Expression Signatures". BMC Bioinformatics 9 (1): 258. doi:10.1186/1471-2105-9-258.