---
title: "Connectivity scores"
author: "Pieter Moris"
date: "21 februari 2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,cache = TRUE)
```

## Unraveling the connection between chemical structure and gene expression.

The aim of this analysis is to investigate the association between the structure of chemical compounds (or their predicted biological activity) and the effects they induce in a gene expression assay. We will group compounds with a similar structure (*chemical fingerprint*) or predicted bioactivity (*target prediction*) using clustering methods. Subsequently the *gene expression profiles* of the compounds within a cluster will be contrasted with each other using a connectivity score approach based on the work of [Lamb et al. (2006)](#lamb) and [Zhang & Gant (2008)](#zang) on the [Connectivity Map](https://www.broadinstitute.org/cmap/). 

Their methods enable us to contrast a number of query expression profiles with a set of reference profiles and assess the *connectivity* between compounds in terms of which genes are regulated up or down. The connection between the query and reference data can be both positive and negative. In the former case, highly up-regulated genes in one set will also be up-regulated in the other set, and vice versa for down-regulated genes. In the latter case, genes with a high expression in one assay, might be strongly down-regulated in the other one. Both situations imply that the two compounds interfere with the same biological processes. If two sets are weakly connected, there is no correlation or overlap between the top genes of either set.

We will use an extension of these methods that scores the connectivity between gene expression profiles using *multiple factor analysis* (MFA).

*To do: description and origins of dataset + further explanation of MFA*

## Clustering chemical compounds
We will employ hierarchical clustering to group the chemical compounds based on their fingerprints or target predictions. In either case we are dealing with a 0/1 matrix where the rows correspond to different compounds. For the chemical fingerprints the columns indicate the presence or absence of specific chemical structures. Together this sequence of ones and zeros describes the entire molecular scaffolding of the compound. In essence we are dealing with a bit string for each compound.

*To do: add description of bioactivity matrix + describe dimensions of data.*

Let's take a look at the chemical fingerprints first. Here is a sneak peak at the data:

```{r,echo=FALSE}
load("MCF7Data.RData")
fingerprintMat <- fingerprintMat*1 # convert logical matrix to binary integer matrix
head(fingerprintMat[,1:4])
```

To properly cluster objects based on binary attributes, we need to define an adequate measure of similarity The *Tanimoto coefficient* (sometimes called the *Jaccard coefficient*) is often used for this purpose in cheminformatics ([MacCuish and MacCuish, 2011](#chem)). It contrasts two objects *x* and *y* of dimension *k* in terms of the number of common attributes, *c*, with the number of attributes unique to either object, *a* and *b*. $$s = \frac{c}{a+b-c} = \frac{\sum_{i=1}^k x_i y_i}{\sum_{i=1}^k a_i^2}$$.

*To do: difference in definitions of Tanimoto, jaccard and soergel distance!*

***

http://www.sequentix.de/gelquest/help/distance_measures.htm

https://en.wikipedia.org/wiki/Jaccard_index Various forms of functions described as Tanimoto similarity and Tanimoto distance occur in the literature and on the Internet. Most of these are synonyms for Jaccard similarity and Jaccard distance, but some are mathematically different. 

https://books.google.be/books?id=ZDDNBQAAQBAJ&pg=PA41&lpg=PA41&dq=tanimoto+clustering&source=bl&ots=vsLen2ZmS5&sig=N16boAKkB5NWLJjeteC4shM6Brc&hl=en&sa=X&ved=0ahUKEwiaiZvsx4vLAhWBfxoKHf-9DrwQ6AEISzAH#v=onepage&q=tanimoto%20clustering&f=false Soergel distance

https://docs.tibco.com/pub/spotfire/6.0.0-november-2013/userguide-webhelp/hc/hc_tanimoto_coefficient.htm another different formula!

***

Its complement ($1-s$), the *Soergel distance* is what we will use for our clustering approach. This is what the distance matrix of the compounds will look like:

*To do: R's dist function using binary seems to provide almost the same result! Again, difference between jaccard/tanimoto definitions?*

```{r}
tanimoto = function(m){
  S = matrix(0,nrow=dim(m)[1],ncol=dim(m)[1])
  m=as.matrix(m)
  N.C=m %*% t(m)
  N.A=m %*% (1-t(m))
  N.B=(1-m) %*% t(m)
  S=N.C/(N.A+N.B+N.C)
  D = 1 - S
  return(D)
}
dist.fingerprintMat <- tanimoto(fingerprintMat)
head(dist.fingerprintMat[,1:4])

# alternative?
test.dist.fingerprintMat <- as.matrix(dist(fingerprintMat,method = 'binary'))
# test.dist.fingerprintMat == dist.fingerprintMat
```

Now we can go ahead and try to cluster these compounds based on their chemical fingerprints. We will use the `hclust` function provided by `R` which implements a basic hierarchical clustering analysis.

**To do: which agglomeration method should I use? complete is the default option, but I believe Woud mentioned average during the meeting?**
**Agglomerative vs regular hierarchical clustering?**
Other clustering options in R specifically designed for the comparison of chemical structures are described in [ChemmineR](https://www.bioconductor.org/packages/3.3/bioc/vignettes/ChemmineR/inst/doc/ChemmineR.html#clustering-identical-or-very-similar-compounds)

```{r,fig.width=8, fig.height=8}
cluster.fingerprintMat <- hclust(as.dist(dist.fingerprintMat), method = "average")
plot(cluster.fingerprintMat,main='Hierarchical clustering of compounds based on their chemical fingerprints',hang=-1,col = "#487AA1", col.main = "#F38630", col.lab = "#F38630", col.axis = "#7C8071",xlab="Compounds",sub="", lwd = 2)
axis(side = 2, at = seq(0, 400, 100), col = "#F38630", labels = TRUE, lwd = 2)
```

Next, we need to decide on an appropriate number of clusters and prune the tree to this size. For example, let's say we opt for ten clusters in total.

```{r,fig.width=8, fig.height=8}
plot(cluster.fingerprintMat,main='Hierarchical clustering of compounds based on their chemical fingerprints',hang=-1,col = "#487AA1", col.main = "#F38630", col.lab = "#F38630", col.axis = "#7C8071",xlab="Compounds",sub="", lwd = 2)
rect.hclust(cluster.fingerprintMat,k=10)
# define groups using groups <- cutree(hc, k = 3)
```

In practise it is better to define the number of clusters based on the distance between certain groups.

**To do: find appropriate method to decide on cluster size! see: http://www2.stat.unibo.it/montanari/Didattica/Multivariate/CA_lab.pdf **

## Connectivity scoring using multiple factor analysis

Now that we have identified clusters of compounds that are similar in their structure (chemical fingerprint), we will investigate whether or not they are also similar according to the gene expression profiles they induce *in vitro*. To do this, we will use MFA to compare the gene expression profile of a certain compound (the query) with the profiles of all the other compounds in its cluster (the reference database). This will be done once for each compound in a cluster, for every cluster we defined earlier. 

**Is this correct? Will I always be comparing 1 query with a reference dataset of size x (being the cluster)?**
**Or should I use an entire cluster as a reference database and all the 56 compounds together as a query dataset? Or vice versa?**

Afterwards, it might also be interesting to contrast entire clusters with one another, i.e. by using one cluster as a query dataset and **another one as the reference dataset/all the other compounds as the reference?**

## Promising reading materials

- [Use of chemical similarity in drug discovery](http://mcc.irb.hr/mcc_04/presentations/butina_d_mcc04_2.pdf)
- [ChemmineR](https://www.bioconductor.org/packages/3.3/bioc/vignettes/ChemmineR/inst/doc/ChemmineR.html#introduction) and [tutorial](http://chemmine.ucr.edu/help/): 
    + contains modules for similarity searching of chemical compounds (Tanimoto coefficient)
    + clustering of compounds by structural and physicochemical similarities is a powerful approach for correlating structural features of compounds with their activities
- [textbook on clustering in drug discovery](https://books.google.be/books?id=ZDDNBQAAQBAJ&pg=PA41&lpg=PA41&dq=tanimoto+clustering&source=bl&ots=vsLen2ZmS5&sig=N16boAKkB5NWLJjeteC4shM6Brc&hl=en&sa=X&ved=0ahUKEwiaiZvsx4vLAhWBfxoKHf-9DrwQ6AEISzAH#v=onepage&q=tanimoto%20clustering&f=false)
- Some cross-validated links:
    + [frequent item set better than clustering?](https://stats.stackexchange.com/questions/86318/clustering-a-binary-matrix)
    + [ordinal vs nominal binary](https://stats.stackexchange.com/questions/116856/hierarchical-or-two-step-cluster-analysis-for-binary-data?rq=1) does the simultaneous absence of both attributes convey any information?
    + [hierarchical clustering](https://stats.stackexchange.com/questions/2717/clustering-with-a-distance-matrix?rq=1)
    + [more clustering options for boolean](https://stats.stackexchange.com/questions/70113/cluster-large-boolean-dataset?rq=1)
- References for clustering:
    + http://handsondatascience.com/ClustersO.pdf

## References
<a name="lamb"></a>- Lamb, J. 2006. "The Connectivity Map: Using Gene-Expression Signatures To Connect Small Molecules, Genes, And Disease". Science 313 (5795): 1929-1935. doi:10.1126/science.1132939.

<a name="chem"></a>- MacCuish, John D, and Norah E MacCuish. 2011. Clustering In Bioinformatics And Drug Discovery. Boca Raton: Taylor & Francis.

<a name="zang"></a> - Zhang, Shu-Dong, and Timothy W Gant. 2008. "A Simple And Robust Method For Connecting Small-Molecule Drugs Using Gene-Expression Signatures". BMC Bioinformatics 9 (1): 258. doi:10.1186/1471-2105-9-258.

